tools/github_update.py

```diff
@@ -96,118 +105,237 @@ def _get_repo_path(self):
+                message = '[{}] {}'.format(datetime.timedelta(seconds=round(time.time()) - start_time), message)
+                os.remove(start_time_path)
+            except:
+                message = '[?] {}'.format(message)
+
+        self._api.set_commit_status(context_label, state, message, details_url)
+
+
+def _get_details_link_url():
+    '''returns the url to be included as the details link in the status'''
+    details_url = os.environ.get('GITHUB_COMMIT_STATUS_URL', '') # custom URL via env
+    if not details_url:
+        details_url = os.environ.get('BUILD_URL', '') # provided by jenkins
+        if details_url:
+            details_url += 'console'
+    return details_url
+
+
+def _should_access_github():
+    if 'WORKSPACE' not in os.environ:
+        # not running in CI. skip actually sending anything to GitHub
+        return False
+    if os.environ.get('GITHUB_DISABLE', ''):
+        # environment has notifications disabled. skip actually sending anything to GitHub
+        return False
+    if not (os.environ.get('GITHUB_COMMIT_STATUS_URL') or os.environ.get('BUILD_URL')):
+        # CI job was not triggered by a PR
+        return False
+    return True
 
 
 def print_help(argv):
-    logger.info('Syntax: {} <state: pending|success|error|failure> <context_label> <status message>'.format(argv[0]))
+    logger.info('Syntax:')
+    logger.info('- Update: {} <state: {}> <context_label> [a status message ...]'.format(
+        argv[0], '|'.join(VALID_STATES)))
+    logger.info('- Reset pending: {} reset [a replacement message ...]'.format(argv[0]))
+
+
+def reset_states(updater, message):
+    if not _should_access_github():
+        # reset disabled due to local build, exit silently
+        return 0
+
+    contexts = [context for context in updater.list_contexts() if context not in BLACKLISTED_CONTEXT_LABELS]
+    if not contexts:
+        # nothing to reset, exit silently
+        return 0
+
+    contexts.sort()
+    logger.info('[GH-STATUS] Resetting {} statuses: {}'.format(len(contexts), ', '.join(contexts)))
+    for context_label in contexts:
+        updater.update(PENDING_STATE, message=message, context_label=context_label)
+    return 0
+
+
+def set_state(updater, state, context_label, message):
+    if context_label in BLACKLISTED_CONTEXT_LABELS:
+        logger.error('Requested context label is on the blacklisted list: {}'.format(BLACKLISTED_CONTEXT_LABELS))
+        return 1
+
+    details_url = _get_details_link_url()
+    if details_url:
+        logmsg = '{} {}: {} ({})'.format(context_label, state, message, details_url)
+    else:
+        logmsg = '{} {}: {}'.format(context_label, state, message)
+
+    if _should_access_github():
+        logger.info('[GH-STATUS] {}'.format(logmsg))
+        updater.update(state, message=message, context_label=context_label, details_url=details_url)
+    else:
+        logger.info('[STATUS] {}'.format(logmsg))
+    return 0
 
 
 def main(argv):
```

Comment:
I don't think it communicates anything to the local user that they don't already know, but not a big deal. not a blocker.